compiler/src/dotty/tools/dotc/transform/CheckShadowing.scala

```diff
@@ -0,0 +1,312 @@
+    def computeTypeParamShadowsFor(parent: Symbol)(using Context): Unit =
+        typeParamCandidates(parent).foreach(typeDef => {
+          val sym = typeDef.symbol
+          val shadowedType =
+            lookForRootShadowedType(sym)
+              .orElse(lookForImportedShadowedType(sym))
+              .orElse(lookForUnitShadowedType(sym))
+          shadowedType.foreach(shadowed =>
+            if !renamedImports.exists(_.contains(shadowed.name.toSimpleName)) then
+              typeParamShadowWarnings += TypeParamShadowWarning(typeDef.srcPos, typeDef.symbol, parent, shadowed)
+          )
+        })
+
+    private def lookForRootShadowedType(symbol: Symbol)(using Context): Option[Symbol] =
+      rootImports.find(p => p.name.toSimpleName == symbol.name.toSimpleName).map(_.symbol)
+
+    private def lookForImportedShadowedType(symbol: Symbol)(using Context): Option[Symbol] =
+      explicitsImports
+        .flatMap(_.flatMap(imp => symbol.isInImport(imp)))
+        .headOption
+
+    private def lookForUnitShadowedType(symbol: Symbol)(using Context): Option[Symbol] =
+      if !ctx.owner.exists then
+        None
+      else
+        val declarationScope = ctx.effectiveScope
+        val res = declarationScope.lookup(symbol.name)
+        res match
+          case s: Symbol if s.isType => Some(s)
+          case _ => lookForUnitShadowedType(symbol)(using ctx.outer)
+
+    /** Register if the valDef is a private declaration that shadows an inherited field */
+    def registerPrivateShadows(valDef: tpd.ValDef)(using Context): Unit =
+      lookForShadowedField(valDef.symbol).foreach(shadowedField =>
+        privateShadowWarnings += PrivateShadowWarning(valDef.startPos, valDef.symbol, shadowedField)
+      )
+
+    private def lookForShadowedField(symDecl: Symbol)(using Context): Option[Symbol] =
+      if symDecl.isPrivate then
+        val symDeclType = symDecl.info
+        val bClasses = symDecl.owner.info.baseClasses
+        bClasses match
+          case _ :: inherited =>
+            inherited
+              .map(classSymbol => symDecl.denot.matchingDecl(classSymbol, symDeclType))
+              .find(sym => sym.name == symDecl.name)
+          case Nil =>
+            None
+      else
+        None
+
+    /** Get the shadowing analysis's result */
+    def getShadowingResult(using Context): ShadowResult =
+      val privateWarnings: List[ShadowWarning] =
+        if ctx.settings.XlintHas.privateShadow then
+          privateShadowWarnings.toList
+        else
+          Nil
+      val typeParamWarnings: List[ShadowWarning] =
+        if ctx.settings.XlintHas.typeParameterShadow then
+          typeParamShadowWarnings.toList
+        else
+          Nil
+      ShadowResult(privateWarnings ++ typeParamWarnings)
+
+    extension (sym: Symbol)
+      /** Given an import and accessibility, return the import's symbol that matches import<->this symbol */
+      private def isInImport(imp: tpd.Import)(using Context): Option[Symbol] =
```

Comment:
Could you please add a short description of how different it's from the extension method in CheckUnused?

Reply:
Yes, I changed it on my last commits, even the method name was not appropriate.